sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,833 @@
+            # on Fortran arrays.
+            X_binned_val = np.ascontiguousarray(X_binned_val)
+            X_binned_train = np.asfortranarray(X_binned_train)
+        else:
+            X_binned_train, y_train = X_binned, y
+            X_binned_val, y_val = None, None
+
+        if self.verbose:
+            print("Fitting gradient boosted rounds:")
+
+        # initialize raw_predictions: those are the accumulated values
+        # predicted by the trees for the training data. raw_predictions has
+        # shape (n_trees_per_iteration, n_samples) where
+        # n_trees_per_iterations is n_classes in multiclass classification,
+        # else 1.
+        n_samples = X_binned_train.shape[0]
+        self._baseline_prediction = self.loss_.get_baseline_prediction(
+            y_train, self.n_trees_per_iteration_
+        )
+        raw_predictions = np.zeros(
+            shape=(self.n_trees_per_iteration_, n_samples),
+            dtype=self._baseline_prediction.dtype
+        )
+        raw_predictions += self._baseline_prediction
+
+        # initialize gradients and hessians (empty arrays).
+        # shape = (n_trees_per_iteration, n_samples).
+        gradients, hessians = self.loss_.init_gradients_and_hessians(
+            n_samples=n_samples,
+            prediction_dim=self.n_trees_per_iteration_
+        )
+
+        # predictors is a matrix (list of lists) of TreePredictor objects
+        # with shape (n_iter_, n_trees_per_iteration)
+        self._predictors = predictors = []
+
+        # Initialize structures and attributes related to early stopping
+        self.scorer_ = None  # set if scoring != loss
+        raw_predictions_val = None  # set if scoring == loss and use val
+        self.train_score_ = []
+        self.validation_score_ = []
+        if self.do_early_stopping_:
+            # populate train_score and validation_score with the predictions
+            # of the initial model (before the first tree)
+
+            if self.scoring == 'loss':
+                # we're going to compute scoring w.r.t the loss. As losses
+                # take raw predictions as input (unlike the scorers), we can
+                # optimize a bit and avoid repeating computing the predictions
+                # of the previous trees. We'll re-use raw_predictions (as it's
+                # needed for training anyway) for evaluating the training
+                # loss, and create raw_predictions_val for storing the
+                # raw predictions of the validation data.
+
+                if self._use_validation_data:
+                    raw_predictions_val = np.zeros(
+                        shape=(self.n_trees_per_iteration_,
+                               X_binned_val.shape[0]),
+                        dtype=self._baseline_prediction.dtype
+                    )
+
+                    raw_predictions_val += self._baseline_prediction
+
+                self._check_early_stopping_loss(raw_predictions, y_train,
```

Comment:
did you ever think about a way to simplify code here and move the logic of checking `loss` vs. `scorer` within the called function? Just asking because it might be difficult.

Reply:
Yes, I had it this way before, with a single `_check_early_stopping` method. It makes the code harder to read IMO since you have to pass it lots of variables that end up not being used depending on whether you're checking the loss or the scorer. So I chose this option which slightly duplicates some logic, but I find it clearer after all.

The thing is, we have a 2-way dichotomy here:

- whether we compute scores based on the loss or based on the scorer (both would take different parameters)
- whther we compute scores on both validation and training data, or just training data.

I agree this is not beautiful, but this is the best I could come up with so far

Reply:
OK so it is fine with me if you also tried stuff before. If it does not
make it simpler then no need to modify it.